### PR TITLE
Fixed passing of user-defined ref props and added them to typescript definitions

### PR DESCRIPTION
**Pull request by cromefire**
- [x] Typescript knows about `MDComponent` on ref elements
- [x] User can pass ref without conflicting with our control ref

### CODE REVIEW

File: ts/Button/index.tsx [hunk 1]
```diff
@@ -35,17 +35,22 @@ export class Button<
...
+  protected materialDom(allprops) {
+    const {ref, className: pcn, ...props} = allprops;
     const ButtonElement = props.href ? 'a' : 'button';
     let className = '';
     this.themeProps.forEach(themeProp => {
       if (themeProp in props && props[themeProp] !== false) {
         className += generateThemeClass(themeProp) + ' ';
       }
     });
+    className += ` ${pcn}`;
```
[the remark this answers is not in the document]
Comment (cromefire) [reply ts/Button/index.tsx]: user-supplied className

[the remark this answers is not in the document]
Comment (cromefire) [reply ts/Button/index.tsx]: PropClassName

**Comment by cromefire** [comment]
But before release could you merge this one?

**Comment by prateekbh** [comment]
i'd like this to be deferred for next months release.

**Comment by cromefire** [comment]
Okay

File: ts/Button/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (prateekbh) [reply ts/Button/index.tsx]: but user supplied class names are added in [materialComponents.ts](https://github.com/prateekbh/preact-material-components/blob/master/ts/Base/MaterialComponent.ts#L68)

**Comment by prateekbh** [comment]
>User can pass ref without conflicting with our control ref

Can you explain? Whats the issue right now?

I am passing ref as a user here https://material.preactjs.com/component/snackbar/ , whats the conflict?

**Comment by cromefire** [comment]
Yes you are right. I thought this will conflict because there is a duplicate ref then? Why does this work, is that some secret (p)react behavior I don't know?

**Comment by cromefire** [comment]
It's like
``` typescript jsx
const props = {ref: (n) => {doSomeStuff(n)}};
return <div ref={(n) => doOtherStuff(n)} {...props} />;
```
wouldn't that conflict?

**Comment by prateekbh** [comment]
cc: @developit can you answer ^

**Comment by cromefire** [comment]
I think I got how ref works now, will redo this

**Pull request closed without merging**